api/src/main/java/org/apache/iceberg/types/ImportCompatibilityChecker.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.types;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Supplier;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class ImportCompatibilityChecker extends CheckCompatibility {
+
+  public ImportCompatibilityChecker(Schema schema, boolean checkOrdering, boolean checkNullability) {
+    super(schema, checkOrdering, checkNullability);
+  }
+
+  @Override
+  public List<String> field(Types.NestedField readField, Supplier<List<String>> fieldErrors) {
+    Types.StructType struct = currentType().asStructType();
+    Types.NestedField field = struct.field(readField.fieldId());
+    List<String> errors = Lists.newArrayList();
+
+    if (field == null) {
+      return NO_ERRORS;
+    }
+    setCurrentType(field.type());
+    try {
+      if (checkNullability() && field.isRequired() && readField.isOptional()) {
+        errors.add(readField.name() + " should be required, but is optional");
+      }
```

Comment:
Correct me if I'm wrong, but this is saying that if the field is marked as optional / nullable in the table to import, but it's marked as required / not-null in the file that's being imported, we'll throw an error when users request nullability checks.

Is there benefit to being that strict? In the case that the target table allows the field to be optional, a file where none of the fields is null violates that constraint (per its data at least).

Though I guess the imported file's own metadata would be incorrect in that case so perhaps it's not overly strict in case we wind up having to rely on the underlying file at some point.

Reply:
It is the other way around.
This case is when table field's schema is `required` and the imported file has nullability as `optional` for the same field.